enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,915 @@
+
+#### Validating a package inside the index
+
+A new sub-command `validate` will be introduced under `opm index`, that will validate the content of each package's json representation. These validations will be in the context of rules about the structure of a package config in an index (required fields, naming conventions etc). A best effort is currently being made to aggregate these rules from various sources and maintained in the [api project's validation library](https://github.com/operator-framework/api/tree/master/pkg/validation). The `opm index validate` will be responsible for validating the index for authors to make sure that the packages in the index conforms to these rules. 
+
+#### Usage of the json representation of packages by olm components
+
+The sqllite database that is currently created inside the index is used to serve contents of the index over a gRPC api.
+
+```go
+rpc ListPackages(ListPackageRequest) returns (stream PackageName) {}
+rpc GetPackage(GetPackageRequest) returns (Package) {}
+rpc GetBundle(GetBundleRequest) returns (Bundle) {}
+rpc GetBundleForChannel(GetBundleInChannelRequest) returns (Bundle) {}
+rpc GetChannelEntriesThatReplace(GetAllReplacementsRequest) returns (stream ChannelEntry) {}
+rpc GetBundleThatReplaces(GetReplacementRequest) returns (Bundle) {}
+rpc GetChannelEntriesThatProvide(GetAllProvidersRequest) returns (stream ChannelEntry) {}
+rpc GetLatestChannelEntriesThatProvide(GetLatestProvidersRequest) returns (stream ChannelEntry) {}
+rpc GetDefaultBundleThatProvides(GetDefaultProviderRequest) returns (Bundle) {}
+rpc ListBundles(ListBundlesRequest) returns (stream Bundle) {}
+```
+
+Once the declarative package configs are available inside an index, these configs will be used to serve the content for these api endpoints instead of the sqllite database. 
+
+
+## Migration Plan
+
+The indexes that exists today have the sqllite database in them, and will not have the package configs for packages that have been added with older versions of `opm` (i.e before `opm index add` creates the package config for the bundle's package). With the release of the new version of the `opm` binary, whenever a new bundle is added to an existing package, or a new package, a config file representing the package will be created inside the index. However, this config file will only contain the bundle/s that's been added with the newer `opm` binary. To prevent a discrepancy between the actual content of the index and what the json representation communicate to a caller of `opm index inspect`, the `inspect` sub-command will need to carry additional logic to first check if the index has an existing sqllite database. If it does, then it will need to convert data in the sqllite database to corresponding package configs for all the packages inside the index. Finally the sqllite database needs to be deleted so that when the index is `inspect`ed again, the configs are presented to the command directly. 
```

Comment:
If they were extracting the db and making changes to it directly, i.e instead of `opm index add --from-index`, then yes there has to be a switch from working with the raw db to the raw json files. 

I agree that's a change, but I'm not sure what you mean exactly when you say `substantial impact`. With the new `opm` version, the equivalent workflow would be: 

1. extract current json representation (instead of DB) from catalog image
2. run opm add against the json representation (instead of DB) to add a new bundle
3. run `opm index update` to push updated json representation to the existing catalog image (instead of rebuilding the catalog image)

Reply:
yeah ok, i think that can work to mimic the current flow.  will still need to think about how the pipeline allows the operator teams to provide their own override json in the future.